adapters/adnuntius/adnuntius_utils.go

```diff
@@ -0,0 +1,202 @@
+	headers.Add("Accept", "application/json")
+	if ortbRequest.Device != nil {
+		if ortbRequest.Device.IP != "" {
+			headers.Add("X-Forwarded-For", ortbRequest.Device.IP)
+		}
+		if ortbRequest.Device.UA != "" {
+			headers.Add("user-agent", ortbRequest.Device.UA)
+		}
+	}
+	return headers
+}
+
+func makeEndpointUrl(ortbRequest openrtb2.BidRequest, a *adapter, noCookies bool) (string, []error) {
+	uri, err := url.Parse(a.endpoint)
+	endpointUrl := a.endpoint
+	if err != nil {
+		return "", []error{fmt.Errorf("failed to parse Adnuntius endpoint: %v", err)}
+	}
+
+	gdpr, consent, err := getGDPR(&ortbRequest)
+	if err != nil {
+		return "", []error{fmt.Errorf("failed to parse Adnuntius endpoint: %v", err)}
+	}
+
+	if !noCookies {
+		var deviceExt extDeviceAdnuntius
+		if ortbRequest.Device != nil && ortbRequest.Device.Ext != nil {
+			if err := jsonutil.Unmarshal(ortbRequest.Device.Ext, &deviceExt); err != nil {
+				return "", []error{fmt.Errorf("failed to parse Adnuntius endpoint: %v", err)}
+			}
+		}
+
+		if deviceExt.NoCookies {
+			noCookies = true
+		}
+	}
+
+	_, offset := a.time.Now().Zone()
+	tzo := -offset / minutesInHour
+
+	q := uri.Query()
+	if gdpr != "" {
+		endpointUrl = a.extraInfo
+		q.Set("gdpr", gdpr)
+	}
+
+	if consent != "" {
+		q.Set("consentString", consent)
+	}
+
+	if noCookies {
+		q.Set("noCookies", "true")
+	}
+
+	q.Set("tzo", fmt.Sprint(tzo))
+	q.Set("format", "prebidServer")
+
+	url := endpointUrl + "?" + q.Encode()
+	return url, nil
+}
+
+func getImpSizes(imp openrtb2.Imp, bidType string) [][]int64 {
+	if bidType == "banner" {
+		if imp.Banner != nil {
+			if len(imp.Banner.Format) > 0 {
+				sizes := make([][]int64, len(imp.Banner.Format))
+				for i, format := range imp.Banner.Format {
+					sizes[i] = []int64{format.W, format.H}
+				}
+
+				return sizes
+			} else if imp.Banner.W != nil && imp.Banner.H != nil {
+				size := make([][]int64, 1)
+				size[0] = []int64{*imp.Banner.W, *imp.Banner.H}
+				return size
+			}
+		}
+	}
+
+	return nil
+}
+
+func getSiteExtAsKv(request *openrtb2.BidRequest) (siteExt, error) {
+	var extSite siteExt
+	if request.Site != nil && request.Site.Ext != nil {
+		if err := jsonutil.Unmarshal(request.Site.Ext, &extSite); err != nil {
+			return extSite, fmt.Errorf("failed to parse ExtSite in Adnuntius: %v", err)
+		}
+	}
+	return extSite, nil
+}
+
+func getGDPR(request *openrtb2.BidRequest) (string, string, error) {
+
+	gdpr := ""
+	var extRegs openrtb_ext.ExtRegs
+	if request.Regs != nil && request.Regs.Ext != nil {
+		if err := jsonutil.Unmarshal(request.Regs.Ext, &extRegs); err != nil {
+			return "", "", fmt.Errorf("failed to parse ExtRegs in Adnuntius GDPR check: %v", err)
+		}
+		if extRegs.GDPR != nil && (*extRegs.GDPR == 0 || *extRegs.GDPR == 1) {
+			gdpr = strconv.Itoa(int(*extRegs.GDPR))
+		}
+	}
+
+	consent := ""
+	if request.User != nil && request.User.Ext != nil {
+		var extUser openrtb_ext.ExtUser
+		if err := jsonutil.Unmarshal(request.User.Ext, &extUser); err != nil {
```

Comment:
Can we use jsonparser.GetString instead of user.ext unmarshal ?

`consent, err := jsonparser.GetString(request.User.Ext, "consent")`

Reply:
I really tried to make this work, but the problem is that if there's no consent added to the request it will throw an error if I use the GetString, making the adapter return a `no-bid`. And what I have now with the `Unmarshal` works. Can you provide an example of what it might look like?